pkg/stacks/relationship_test.go

```diff
@@ -0,0 +1,310 @@
+*/
+
+package stacks
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	namespace             = "cool-namespace"
+	uidString             = "definitely-a-uuid"
+	resourceName          = "cool-resource"
+	thirtyTwoAs           = strings.Repeat("a", 32)
+	sixtyOneAs            = strings.Repeat("a", 61)
+	sixtyThreeAs          = strings.Repeat("a", 63)
+	truncatedAsAndZ       = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-ukl4i"
+	truncatedSixtyThreeAs = "aaaaaaaaaaaaaaaaaaaaaaaaaa-apyj6"
+)
+
+func resource(namespace, resourceName, uidString string) *corev1.Secret {
+	uid := types.UID(uidString)
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      resourceName,
+			UID:       uid,
+		},
+	}
+}
+func TestParentLabels(t *testing.T) {
+	tests := []struct {
+		name string
+		i    KindlyIdentifier
+		want map[string]string
+	}{
+		{
+			name: "Success",
+			i:    resource(namespace, resourceName, uidString),
+			want: map[string]string{
+				LabelParentNamespace: namespace,
+				LabelParentName:      resourceName,
+				LabelParentUID:       uidString,
+				LabelParentGroup:     "",
+				LabelParentKind:      "",
+				LabelParentVersion:   "",
+			},
+		},
+		{
+			name: "Max",
+			i:    resource(sixtyThreeAs, sixtyThreeAs, uidString),
+			want: map[string]string{
+				LabelParentNamespace: sixtyThreeAs,
+				LabelParentName:      sixtyThreeAs,
+				LabelParentUID:       uidString,
+				LabelParentGroup:     "",
+				LabelParentKind:      "",
+				LabelParentVersion:   "",
+			},
+		},
+		{
+			name: "Truncated",
+			i:    resource(sixtyThreeAs+"z", sixtyThreeAs+"z", uidString),
+			want: map[string]string{
+				LabelParentNamespace: truncatedAsAndZ,
+				LabelParentName:      truncatedAsAndZ,
+				LabelParentUID:       uidString,
+				LabelParentGroup:     "",
+				LabelParentKind:      "",
+				LabelParentVersion:   "",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := ParentLabels(tt.i)
+			if diff := cmp.Diff(tt.want, got); diff != "" {
+				t.Errorf("ParentLabels() -want, +got:\n%v", diff)
+			}
+		})
+	}
+}
+
+func TestMultiParentLabel(t *testing.T) {
+
+	tests := []struct {
+		name        string
+		stackParent metav1.Object
+		want        string
+	}{
+		{
+			name:        "Simple",
+			stackParent: resource(namespace, resourceName, uidString),
+			want:        "parent.stacks.crossplane.io/cool-namespace-cool-resource",
+		},
+		{
+			name:        "NotTruncatedNS",
+			stackParent: resource(thirtyTwoAs, "z", uidString),
+			want:        "parent.stacks.crossplane.io/" + thirtyTwoAs + "-z",
+		},
+		{
+			name:        "TruncatedNS",
+			stackParent: resource(sixtyThreeAs+"z", resourceName, uidString),
+			// notice ukl4i is also the suffix with 64 a's because the sum is
```

Comment:
I didn't follow that - we can get the same suffix for different strings?

Reply:
I'll revisit this comment.  The intent was to express the following:

If the base32 sum of a digest of  "aaaaaaaaaaa" is "foo" (with a suffix length of 4).

Truncating this string to a length of 11 would result in "aaaaaaaaaaa"
Truncating this string to a length of 8 would result in "aaaa-foo"
Truncating this string to a length of 7 would result in "aaa-foo"
Truncating this string to a length of 5 would result in "a-foo"
 
"aaaaaaaaaaa"